source/common/protobuf/utility.cc

```diff
@@ -586,6 +590,148 @@ std::string MessageUtil::CodeEnumToString(ProtobufUtil::error::Code code) {
+}
+
+// Forward declaration for mutually-recursive helper functions.
+void redact(Protobuf::Message* message, bool ancestor_is_sensitive);
+
+// To redact an `Any`, we have to unpack it to its original type to redact it.
+bool redactAny(Protobuf::Message* message, bool ancestor_is_sensitive) {
+  auto* any = dynamic_cast<ProtobufWkt::Any*>(message);
+  if (any != nullptr) {
+    Protobuf::DynamicMessageFactory message_factory;
+    auto typed_message = typeUrlToNewMessage(message_factory, any->type_url());
+    if (typed_message == nullptr) {
+      // If the type URL doesn't correspond to a known proto, give up redacting and treat this
+      // message the same as a `ProtobufWkt::Struct`. See the documented limitation on
+      // `MessageUtil::redact()` for more context.
+      ENVOY_LOG_MISC(warn, "Could not redact ProtobufWkt::Any with unknown type URL {}",
+                     any->type_url());
+      return false;
+    }
+
+    any->UnpackTo(typed_message.get());
+    redact(typed_message.get(), ancestor_is_sensitive);
+    any->PackFrom(*typed_message);
+    return true;
+  }
+  return false;
+}
+
+// To redact a `TypedStruct`, we have to reify it based on its `type_url` to redact it.
+bool redactTypedStruct(Protobuf::Message* message, bool ancestor_is_sensitive) {
+  auto* typed_struct = dynamic_cast<udpa::type::v1::TypedStruct*>(message);
+  if (typed_struct != nullptr) {
+    Protobuf::DynamicMessageFactory message_factory;
+    auto typed_message = typeUrlToNewMessage(message_factory, typed_struct->type_url());
+    if (typed_message == nullptr) {
+      // If the type URL doesn't correspond to a known proto, give up redacting and treat this
+      // message the same as a `ProtobufWkt::Struct`. See the documented limitation on
+      // `MessageUtil::redact()` for more context.
+      ENVOY_LOG_MISC(warn, "Could not redact udpa::type::v1::TypedStruct with unknown type URL {}",
+                     typed_struct->type_url());
+      return false;
+    }
+
+    MessageUtil::jsonConvert(typed_struct->value(), ProtobufMessage::getNullValidationVisitor(),
+                             *typed_message);
+    redact(typed_message.get(), ancestor_is_sensitive);
+    MessageUtil::jsonConvert(*typed_message, *(typed_struct->mutable_value()));
+    return true;
+  }
+  return false;
+}
+
+// `DataSource`, which is annotated as `sensitive` in `TlsCertificate` and `TlsSessionTicketKeys`
+// requires special handling.
+bool redactDataSource(Protobuf::Message* message, bool ancestor_is_sensitive) {
+  auto* data_source = dynamic_cast<envoy::config::core::v3alpha::DataSource*>(message);
+  if (data_source != nullptr) {
+    if (ancestor_is_sensitive) {
+      switch (data_source->specifier_case()) {
+      case envoy::config::core::v3alpha::DataSource::SPECIFIER_NOT_SET:
+        // If the data source is empty, no work is needed.
+        break;
+      case envoy::config::core::v3alpha::DataSource::kFilename:
```

Comment:
either not redacted or making it as "[redacted]" is fine. (if later makes us think less specific for DataSource)

Reply:
Ok! If I can also eliminate the special-case handling for inline bytes (below) then I'll get rid of this. If not, I'll just keep it.